src/darwin/Framework/CHIP/CHIPIMDispatch.mm

```diff
@@ -0,0 +1,284 @@
+ *    limitations under the License.
+ */
+#import <Foundation/Foundation.h>
+
+#include <access/SubjectDescriptor.h>
+#include <app-common/zap-generated/att-storage.h>
+#include <app-common/zap-generated/callback.h>
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app-common/zap-generated/ids/Attributes.h>
+#include <app-common/zap-generated/ids/Clusters.h>
+#include <app-common/zap-generated/ids/Commands.h>
+#include <app/AttributeAccessInterface.h>
+#include <app/CommandHandler.h>
+#include <app/ConcreteAttributePath.h>
+#include <app/ConcreteCommandPath.h>
+#include <app/MessageDef/AttributeReportIBs.h>
+#include <app/MessageDef/StatusIB.h>
+#include <app/WriteHandler.h>
+#include <app/data-model/Decode.h>
+#include <lib/core/CHIPError.h>
+#include <lib/core/CHIPTLV.h>
+#include <lib/core/DataModelTypes.h>
+#include <lib/core/Optional.h>
+#include <protocols/interaction_model/Constants.h>
+
+/**
+ * This file defines the APIs needed to handle interaction model dispatch.
+ * These are the APIs normally defined in
+ * src/app/util/ember-compatibility-functions.cpp and the generated
+ * IMClusterCommandHandler.cpp but we want a different implementation of these
+ * to enable more dynamic behavior, since not all framework consumers will be
+ * implementing the same server clusters.
+ */
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+
+namespace {
+
+// TODO: Maybe consider making this configurable?
+constexpr EndpointId kSupportedEndpoint = 0;
+
+} // anonymous namespace
+
+namespace chip {
+namespace app {
+
+    using Protocols::InteractionModel::Status;
+    using Access::SubjectDescriptor;
+
+    namespace {
+
+        Status DetermineAttributeStatus(const ConcreteAttributePath & aPath, bool aIsWrite)
+        {
+            // We don't have any non-global attributes.
+            using namespace Globals::Attributes;
+
+            // TODO: Consider making this configurable for applications that are not
+            // trying to be an OTA provider, though in practice it just affects which
+            // error is returned.
+            if (aPath.mEndpointId != kSupportedEndpoint) {
+                return Status::UnsupportedEndpoint;
+            }
+
+            // TODO: Consider making this configurable for applications that are not
+            // trying to be an OTA provider, though in practice it just affects which
+            // error is returned.
+            if (aPath.mClusterId != OtaSoftwareUpdateProvider::Id) {
+                return Status::UnsupportedCluster;
+            }
+
+            switch (aPath.mAttributeId) {
+            case AttributeList::Id:
+                FALLTHROUGH;
+            case ClientGeneratedCommandList::Id:
+                FALLTHROUGH;
+            case ServerGeneratedCommandList::Id:
+                FALLTHROUGH;
+                // When EventList is supported, include it here.
+#if 0
+    case EventList::Id:
+        FALLTHROUGH;
+#endif
```

Comment:
Is this needed?

Reply:
We could leave it out; the trick is not forgetting to add it later....